proof/access-control/AARCH64/ArchFinalise_AC.thy

```diff
@@ -0,0 +1,424 @@
+  for pas_refined[wp]: "pas_refined aag"
+
+lemma set_vcpu_state_vrefs[wp]:
+  "set_vcpu ptr vcpu \<lbrace>\<lambda>s. P (state_vrefs s)\<rbrace>"
+  apply (wpsimp wp: set_vcpu_wp)
+  apply (erule_tac P=P in rsubst)
+  apply (fastforce intro: state_vrefs_eqI simp: opt_map_def typ_at_eq_kheap_obj)
+  done
+
+lemma state_vrefs_set_asid_pool_vmid:
+  assumes "pool_for_asid asid s = Some pool_ptr"
+      and "asid_pools_of s pool_ptr = Some pool"
+      and "pool (asid_low_bits_of asid) = Some entry"
+    shows "state_vrefs
+            (s\<lparr>kheap := (kheap s)
+                 (pool_ptr \<mapsto>
+                  ArchObj
+                   (ASIDPool
+                     (\<lambda>a. if a = asid_low_bits_of asid
+                           then Some (ASIDPoolVSpace vmid (ap_vspace entry)) else pool a)))\<rparr>)
+            x
+           \<subseteq> state_vrefs s x"
+  (is "state_vrefs ?s' _ \<subseteq> state_vrefs _ _")
+  apply (rule state_vrefs_subseteq)
+  using assms
+  by (auto simp: vspace_for_pool_def entry_for_pool_def opt_map_def obind_def obj_at_def
+          split: option.splits)
+
+lemma update_asid_pool_entry_vmid_pas_refined[wp]:
+  "update_asid_pool_entry (\<lambda>entry. Some (ASIDPoolVSpace vmid (ap_vspace entry))) asid \<lbrace>pas_refined aag\<rbrace>"
+  unfolding update_asid_pool_entry_def set_asid_pool_def
+  apply (wpsimp wp: set_object_wp)+
+  apply (erule pas_refined_subseteq; clarsimp?)
+     apply (rule caps_of_state_fun_upd)
+     apply (clarsimp simp: obj_at_def opt_map_def split: option.splits)
+    apply (erule rev_subsetD, rule state_vrefs_subseteq)
+       apply (auto simp: vspace_for_pool_def entry_for_pool_def opt_map_def obind_def
+                  split: option.splits kernel_object.splits)[4]
+   apply (rule thread_st_auth_fun_upd)
+   apply (clarsimp simp: obj_at_def asid_pools_of_ko_at get_tcb_def)
+  apply (rule thread_bound_ntfns_fun_upd)
+  apply (clarsimp simp: asid_pools_of_ko_at get_tcb_def obj_at_def)
+  done
+
+lemma tcb_vcpu_update_pas_refined[wp]:
+  "arch_thread_set v t \<lbrace>pas_refined aag\<rbrace>"
+  apply (wpsimp wp: arch_thread_set_wp)
+  apply (erule pas_refined_subseteq; clarsimp?)
+     apply (rule caps_of_state_fun_upd)
+     apply (clarsimp simp: obj_at_def get_tcb_def tcb_cap_cases_def
+                    split: option.splits kernel_object.splits)
+    apply (erule rev_subsetD, rule state_vrefs_subseteq)
+       apply (auto simp: vspace_for_pool_def entry_for_pool_def opt_map_def obind_def get_tcb_def
+                  split: option.splits kernel_object.splits)[4]
+   apply (rule thread_st_auth_fun_upd)
+   apply (clarsimp simp: obj_at_def asid_pools_of_ko_at get_tcb_def)
+  apply (rule thread_bound_ntfns_fun_upd)
+  apply (clarsimp simp: obj_at_def asid_pools_of_ko_at get_tcb_def)
+  done
+
+lemma set_vcpu_pas_refined:
+  "\<lbrace>\<lambda>s. pas_refined aag s \<and> vcpu_at ptr s\<rbrace>
+   set_vcpu ptr vcpu
+   \<lbrace>\<lambda>_ s. pas_refined aag s\<rbrace>"
```

Comment:
This is modelled on older lemmas I assume, because we did this a lot in the beginning -- the stronger invariant form `set_vcpu ptr vcpu ⦃pas_refined aag⦄` should also be true, because `set_vcpu` should fail if there is no vcpu, and therefore partial correctness triples shouldn't need the `vcpu_at`. The stronger form is easier to use.